src/lib.rs

```diff
@@ -198,6 +204,66 @@ impl OpenOptions {
     }
 }
 
+/// Iterate over the contents of a directory. Created by calling read_dir() on
+/// an opened directory. Each item yielded by the iterator is an io::Result to
+/// allow communication of io errors as the iterator is advanced.
+///
+/// To the greatest extent possible the underlying OS semantics are preserved.
+/// That means that `.` and `..` entries are exposed, and that no sort order is
+/// guaranteed by the iterator.
+#[derive(Debug)]
+pub struct ReadDir<'a> {
+    _impl: ReadDirImpl<'a>,
+}
+
+impl<'a> ReadDir<'a> {
+    pub fn new(d: &'a mut File) -> Result<Self> {
+        Ok(ReadDir {
+            _impl: ReadDirImpl::new(d)?,
+        })
+    }
+}
+
+impl Iterator for ReadDir<'_> {
+    type Item = Result<DirEntry>;
+
+    fn next(&mut self) -> Option<Result<DirEntry>> {
+        self._impl
+            .next()
+            .map(|entry| entry.map(|_impl| DirEntry { _impl }))
+    }
+}
+
+/// The returned type for each entry found by [`read_dir`].
+///
+/// Each entry represents a single entry inside the directory. Platforms that
+/// provide rich metadata may in future expose this through methods or extension
+/// traits on DirEntry.
+///
+/// For now however, only the [`name()`] is exposed. This does not imply any
```

Comment:
I see your point about it being racy, but, does `open_at` give you a way to avoid the race consistently.

If there's an entry that's flipping between being a symlink, directory, file, and something else, and I want to readlink, readlink, read, or error out... I'd need a `nofollow` option to start with, and then I guess the application code needs to retry various options until something succeeds? Seems a bit messy?

(Probably not a reason to block addition of this with just the name to start with, just a thought.)

Reply:
> Something like [sourcefrog@88faff1](https://github.com/sourcefrog/fs_at/commit/88faff1838d536b0fe77d982216e31b9ca11c4d4) works on Linux at least.

Thats ... fascinating. What bakes my brain here is that this is a *new fd* - where is the state coming from? (The previous iteration over the DIR* contents should be wiped out when fdopendir is called : https://docs.rs/nix/0.24.2/src/nix/dir.rs.html#57

And we never actually read from the original fd: the pattern is:
- open dir (fd 3)
- fcntl(3, F_DUPFD_CLOEXEC, 0) = 4
- fdopendir(4) = DIR*
- drop for `Dir` calls closedir correctly., which closes fd 4.

It works on Linux here outside docker. Doesn't work within docker. That means we don't require a kernel version change to trigger the situation.

Linux tr2vm 5.13.0-52-generic #59-Ubuntu SMP Wed Jun 15 20:17:13 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
libc6:amd64                        2.34-0ubuntu3.2

The failing docker container contents:
ii  libc6:amd64    2.35-0ubuntu3.1 amd64        GNU C Library: Shared libraries

Reply:
> I see your point about it being racy, but, does `open_at` give you a way to avoid the race consistently.
> 
> If there's an entry that's flipping between being a symlink, directory, file, and something else, and I want to readlink, readlink, read, or error out... I'd need a `nofollow` option to start with, and then I guess the application code needs to retry various options until something succeeds? Seems a bit messy?
> 
> (Probably not a reason to block addition of this with just the name to start with, just a thought.)

NoFollow is absolutely needed yes. We also need an inode abstraction (to permit safeish `open_at(child, '..')`) to walk back up the tree. But then you don't need to retry, you just open, interrogate, and if a symlink follow the pointer from userspace, otherwise process the object you obtained from open_at.

Reply:
> > Something like [sourcefrog@88faff1](https://github.com/sourcefrog/fs_at/commit/88faff1838d536b0fe77d982216e31b9ca11c4d4) works on Linux at least.
> 
> Thats ... fascinating. What bakes my brain here is that this is a _new fd_ - where is the state coming from? (The previous iteration over the DIR* contents should be wiped out when fdopendir is called : https://docs.rs/nix/0.24.2/src/nix/dir.rs.html#57
> 
> And we never actually read from the original fd: the pattern is:
> 
> * open dir (fd 3)
> * fcntl(3, F_DUPFD_CLOEXEC, 0) = 4
> * fdopendir(4) = DIR*
> * drop for `Dir` calls closedir correctly., which closes fd 4.
> 
> It works on Linux here outside docker. Doesn't work within docker. That means we don't require a kernel version change to trigger the situation.
> 
> Linux tr2vm 5.13.0-52-generic #59-Ubuntu SMP Wed Jun 15 20:17:13 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux libc6:amd64 2.34-0ubuntu3.2
> 
> The failing docker container contents: ii libc6:amd64 2.35-0ubuntu3.1 amd64 GNU C Library: Shared libraries

Installed an impish container (same as host os that 'works') and it still fails - so this is looking very specific to docker.

Going to try a couple things :). Also I've realised nix has poor readdir hygiene, filing a bug there.

Reply:
https://github.com/nix-rust/nix/issues/1783

Reply:
This leaves a strong hint (from the NetBSD man page in particular) https://github.com/nix-rust/nix/issues/1784 

So I think seek might be right, though I'm still suspicious of underlying things here.

Reply:
It failed for me first time, outside of Docker, on Pop_OS 22.04 Linux lift 

5.18.10-76051810-generic #202207071639~1659108431~22.04~c9172fb SMP PREEMPT_DYNAMIC Fri J x86_64 x86_64 x86_64 GNU/Linux

I can't immediately cite chapter and verse but I'm not surprised that dup'd file descriptors share a seek position. 

<https://man7.org/linux/man-pages/man2/dup.2.html>

>        The dup() system call allocates a new file descriptor that refers to the same open file description as the descriptor oldfd. 

https://man7.org/linux/man-pages/man2/open.2.html

>         A call to open() creates a new open file description, an entry in
>        the system-wide table of open files.  The open file description
>        records the file offset and the file status flags (see below).  A
>        file descriptor is a reference to an open file description; this
>        reference is unaffected if pathname is subsequently removed or
>        modified to refer to a different file.  For further details on
>        open file descriptions, see NOTES.

So my model was that you have a single open file object in the kernel with two fd numbers pointing to it. If you read from it by one, you have to seek it back in the other.

That does not explain why it would succeed in some cases though!

Reply:
> NoFollow is absolutely needed yes. We also need an inode abstraction (to permit safeish open_at(child, '..')) to walk back up the tree. But then you don't need to retry, you just open, interrogate, and if a symlink follow the pointer from userspace, otherwise process the object you obtained from open_at.

But with `O_NOFOLLOW` you'll fail to open the file. You need to then retry with `readlinkat`.

I split out #13 on this too.

The other case I though of where `statat` will help beyond `fstat` is a file that you don't have permission to read.

Reply:
Yes, the success in some cases is very confusing. We may need to add a note that the same file object is in use and thus the file location offset can be updated by readdir.

O_PATH permits fstat after opening, but the portability concerns are there :/. 

The advice for programs that want to be secure and run on Darwin, from that stackexchange, is to tell folk that they need to grant read access to permit such software to work, or document its vulnerability to races on Darwin.